Add more matprop output capability

GiudGiud opened the pull request:
Needed for making new tests in Pronghorn
@grmnptr or @lindsayad

GiudGiud committed:
Add a vector material property extractor to auxvariables, refs #19382

GiudGiud committed:
Add a function vector functor material, refs #16809

GiudGiud committed:
Add test for capturing a component of a vector matprop in an auxvariable, refs #19382
Clang format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/891017/) on b168957 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19402/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19402/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d3e57ccfba29379e643967ccc89b637b34007001`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/891066/) on 87aebcc wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19402/site)

This comment will be updated on new commits.

GiudGiud committed:
Add documentation for the function vector material, refs #19382

GiudGiud committed:
Add documentation for the vector functor matprop -> auxvariable auxkernel, refs #19382
Change test to use elemental variable, for consistency at the block interface

GiudGiud commented on framework/doc/content/source/auxkernels/FunctorVectorMatPropElementalAux.md:
```diff
@@ -0,0 +1,19 @@
+[!param](/AuxKernels/FunctorVectorMatPropElementalAux/component) parameter.
+
+!alert note
+The version of this material for automatic differentiation (AD) material properties is
```
```suggestion
The version of this auxkernel for automatic differentiation (AD) material properties is
```

lindsayad commented on framework/doc/content/source/auxkernels/FunctorMatPropElementalAux.md:
```diff
@@ -7,6 +7,10 @@
 material property with an element argument.
 
+!alert note
+The version of this material for automatic differentiation (AD) material properties is
```
```suggestion
The version of this auxiliary kernel for automatic differentiation (AD) material properties is
```

lindsayad commented on framework/doc/content/source/materials/GenericConstantVectorFunctorMaterial.md:
```diff
@@ -2,7 +2,7 @@
 !syntax description /Materials/GenericConstantVectorFunctorMaterial
 
-The functor version of [GenericConstantVectorFunctorMaterial.md], this can be
+The functor version of [GenericConstantFunctorMaterial.md], this can be
```
This clearly wasn't right before but it still doesn't look right now. Guessing that `Functor` should not be in the name of the thing we are a functor version of

GiudGiud replied:
oh true true must have misclicked somehow

lindsayad commented on framework/src/auxkernels/FunctorVectorMatPropElementalAux.C:
```diff
@@ -0,0 +1,43 @@
+  params.addClassDescription(
+      "Evaluates a functor vector material property on the current element."
+      "For finite volume, this evaluates the material property at the centroid.");
+  params.addRequiredParam<MaterialPropertyName>("mat_prop", "The functor material property");
```
I started this myself, but we can move away from implying what kind of functor this is, e.g. this aux kernel (well the AD version of it) would work equally well with a functor that is in fact a MOOSE vector variable. Then this parameter type would be our generic `MooseFunctorName`, and the member name could be more general. In my interpolator PR I'll be adding more types of functors and it's hard to anticipate what more we will add in the future. With that in mind, we don't need to restrict the notion of what our inputs are. (Also in the interpolator PR I actually add the ability to change the backing functor through the course of the simulation and at the same time allow consumers to continue to hold a valid reference using the letter-and-envelope idiom)

GiudGiud replied:
yes but as you said we ve kept on the traditional naming (function / matprop / auxvars) so far.

If I drop this here I ll refact all our objects then.

lindsayad replied:
I'm admittedly based on a commit that's a little old but I don't see that many objects that have specific meaning
```
(base) lindad@inl610212:~/projects/moose(insfv-interpolator)$ git grep -n "getFunctor.*mat"
framework/src/auxkernels/FunctorMatPropElementalAux.C:31:  : AuxKernel(parameters), _mat_prop(getFunctor<GenericReal<is_ad>>("mat_prop"))
modules/navier_stokes/src/postprocessors/InternalVolumetricFlowRate.C:52:    _advected_material_property(getFunctor<ADReal>("advected_mat_prop"))
modules/navier_stokes/src/postprocessors/VolumetricFlowRate.C:52:    _advected_material_property(getFunctor<ADReal>("advected_mat_prop"))
test/src/bcs/MatPropBC.C:24:  : ADIntegratedBC(parameters), _coef(getFunctor<ADReal>("mat_prop"))
test/src/fvkernels/FVFunctorElementalKernel.C:25:  : FVElementalKernel(params), _functor_prop(getFunctor<ADReal>("mat_prop_name"))
```
Of the production objects, I would contend that only `FunctorMatPropElementalAux` is poorly named

lindsayad replied:
It's a shame about the parameter names for `(Internal)VolumetricFlowRate` ... if they were just `advected_prop` then I think there's no connotation that this is a material property. Even now, the parameter type for those is `MooseFunctorName`

GiudGiud replied:
there s also the VarMatFunctorAux or something.

GiudGiud replied:
I agree it s not a lot. Which is why I m offering to do it ^^

lindsayad replied:
I couldn't tell whether you were being sassy or not 😆 Hard to tell over a comp. If you're offering, then I think that would be awesome

GiudGiud replied:
ok I m gonna keep the two volumetric flow rate ones for when I work on SAM coupling. I think I ll see more clearly what the needs are then. I suspect we will be able to fuse the advected_variable and advected_matprop into an advected_functor parameter, but i d like to be sure

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/891556/) on fa04433 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>28e774</th>
      <th colspan=4>#19402 fa0443</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>81.95%</td>
      <td>81.96%</td>
      <td>+0.01%</td>
      <td>94.87%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>72274</td>
      <td>72311</td>
      <td>+37</td>
      <td>37</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15915</td>
      <td>15916</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19402/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19402/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/tensor_mechanics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19402/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Use functors for object naming more

GiudGiud committed:
Fixup NS test failures
Clang formatting

GiudGiud commented:
Aside from the test failure (something with the caching, not sure I understand it yet), and possibly breaking Pgh, this should be good for review.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/892689/) on 341fd55 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19402/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19402/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 44d8c685e45727d1bec41d4a56196a157d1a5e58`

GiudGiud committed:
Finish renaming FVPropValPerSubdomainMat to PiecewiseConstantByBlockMaterial, refs #16809
Fixup tests requirements since renaming of design files
Add execute_on nonlinear to functor materials, as it matters for variable functors
Remove doc file for GenericFunctionFunctorVectorMaterial since consolidated with another

GiudGiud commented:
ok should be ready for review @lindsayad

lindsayad commented on framework/src/materials/GenericFunctorMaterial.C:
```diff
@@ -0,0 +1,84 @@
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params += SetupInterface::validParams();
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_LINEAR, EXEC_NONLINEAR};
```
I probably mislead/misunderstood you earlier. I think the default for this should be `EXEC_ALWAYS`. Especially for constants, this evaluation should be very cheap. By setting `execute_on` to anything other than `EXEC_ALWAYS` we are committing to caching

lindsayad replied:
It looks like you changed this back somehow

GiudGiud replied:
ugh must have been the force-push

lindsayad commented on framework/src/materials/GenericFunctorMaterial.C:
```diff
@@ -0,0 +1,84 @@
+  for (const auto i : make_range(_num_props))
+  {
+    auto & prop = declareFunctorProperty<GenericReal<is_ad>>(_prop_names[i]);
+    prop.setFunctor(
+        _mesh, blockIDs(), [this, i](const auto & r, const auto & t) -> GenericReal<is_ad> {
+          return (*_functors[i])(r, t);
+        });
```
The power of functors! Look at this arbitrary evaluation chaining!!

lindsayad replied:
But now I am wondering ... if we are just forwarding evaluation to another functor, should *this* functor even exist? What new thing is it doing that the forwarded-to functor didn't do?

GiudGiud replied:
it changes the name lol

GiudGiud replied:
and in the future, it may or may not be able to change the AD type

GiudGiud replied:
oh and it can change the caching behavior!

GiudGiud replied:
I m just trying to get my PR in tbh. If you want a deletion i ll do it

lindsayad replied:
Well you're doing the kind act of consolidating ... so I'll let this go for now. At a minimum you did not add any new unnecessary objects

GiudGiud replied:
actually it's cool to do something like that:
```
  [firebrick_properties]
    type = ADGenericFunctorMaterial
    prop_names = 'rho_s         cp_s        k_s'
    prop_values = '${solid_rho} ${solid_cp} 0.26'
    block = '10'
  []
 ```
some kernels and materials in Pgh just like to retrieve the functors directly, not in the parameters but through the PGH namespace

lindsayad commented on test/tests/materials/functor_properties/bc/bc.i:
```diff
@@ -33,9 +33,10 @@
+    type = ADGenericFunctorMaterial
+    prop_names = 'prop'
+    prop_values = 'v'
+    execute_on = 'LINEAR NONLINEAR'
```
Is this necessary now?

GiudGiud replied:
just for awareness

GiudGiud replied:
now it is, since we are always by default

GiudGiud commented on framework/src/materials/GenericVectorFunctorMaterial.C:
```diff
@@ -0,0 +1,79 @@
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params += SetupInterface::validParams();
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_LINEAR, EXEC_NONLINEAR};
```
```suggestion
  params.set<ExecFlagEnum>("execute_on") = {EXEC_ALWAYS};
```

GiudGiud commented on framework/src/materials/GenericFunctorMaterial.C:
```diff
@@ -0,0 +1,84 @@
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params += SetupInterface::validParams();
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_LINEAR, EXEC_NONLINEAR};
```
```suggestion
  params.set<ExecFlagEnum>("execute_on") = {EXEC_ALWAYS};
```

GiudGiud replied:
```suggestion
  params.set<ExecFlagEnum>("execute_on") = {EXEC_ALWAYS};
```

GiudGiud committed:
Update framework/src/materials/GenericFunctorMaterial.C

lindsayad commented on framework/src/materials/GenericVectorFunctorMaterial.C:
```diff
@@ -0,0 +1,79 @@
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params += SetupInterface::validParams();
+  params.set<ExecFlagEnum>("execute_on") = {EXEC_LINEAR, EXEC_NONLINEAR};
```
```suggestion
  params.set<ExecFlagEnum>("execute_on") = {EXEC_ALWAYS};
```

lindsayad replied:
And here

GiudGiud replied:
nice thanks for double checking

lindsayad committed:
Update framework/src/materials/GenericVectorFunctorMaterial.C
